pkg/ccl/bulkccl/desc_rewriting.go

```diff
@@ -0,0 +1,726 @@
+				}
+			} else {
+				// If the database is not multi-region enabled, ensure that we don't
+				// write any multi-region table descriptors into it.
+				if table.GetLocalityConfig() != nil {
+					return pgerror.Newf(pgcode.FeatureNotSupported,
+						"cannot write descriptor for multi-region table %s into non-multi-region database %s",
+						table.GetName(),
+						dbDesc.GetName(),
+					)
+				}
+			}
+
+			if err := descsCol.WriteDescToBatch(
+				ctx, false /* kvTrace */, tables[i].(catalog.MutableDescriptor), b,
+			); err != nil {
+				return err
+			}
+			// Depending on which cluster version we are restoring to, we decide which
+			// namespace table to write the descriptor into. This may cause wrong
+			// behavior if the cluster version is bumped DURING a restore.
+			tkey := catalogkv.MakeObjectNameKey(
+				ctx,
+				settings,
+				table.GetParentID(),
+				table.GetParentSchemaID(),
+				table.GetName(),
+			)
+			b.CPut(tkey.Key(codec), table.GetID(), nil)
+		}
+
+		// Write all type descriptors -- create namespace entries and write to
+		// the system.descriptor table.
+		for i := range types {
+			typ := types[i]
+			updatedPrivileges, err := getRestoringPrivileges(ctx, codec, txn, typ, user, wroteDBs, descCoverage)
+			if err != nil {
+				return err
+			}
+			if updatedPrivileges != nil {
+				if mut, ok := typ.(*typedesc.Mutable); ok {
+					mut.Privileges = updatedPrivileges
+				} else {
+					log.Fatalf(ctx, "wrong type for type %d, %T, expected Mutable",
+						typ.GetID(), typ)
+				}
+			}
+			if err := descsCol.WriteDescToBatch(
+				ctx, false /* kvTrace */, typ.(catalog.MutableDescriptor), b,
+			); err != nil {
+				return err
+			}
+			tkey := catalogkv.MakeObjectNameKey(ctx, settings, typ.GetParentID(), typ.GetParentSchemaID(), typ.GetName())
+			b.CPut(tkey.Key(codec), typ.GetID(), nil)
+		}
+
+		for _, kv := range extra {
+			b.InitPut(kv.Key, &kv.Value, false)
+		}
+		if err := txn.Run(ctx, b); err != nil {
+			if errors.HasType(err, (*roachpb.ConditionFailedError)(nil)) {
+				return pgerror.Newf(pgcode.DuplicateObject, "table already exists")
+			}
+			return err
+		}
+		return nil
+	}()
+	return errors.Wrapf(err, "restoring table desc and namespace entries")
+}
+
+func getRestoringPrivileges(
+	ctx context.Context,
+	codec keys.SQLCodec,
+	txn *kv.Txn,
+	desc catalog.Descriptor,
+	user security.SQLUsername,
+	wroteDBs map[descpb.ID]catalog.DatabaseDescriptor,
+	descCoverage tree.DescriptorCoverage,
+) (updatedPrivileges *descpb.PrivilegeDescriptor, err error) {
+	switch desc := desc.(type) {
+	case catalog.TableDescriptor, catalog.SchemaDescriptor:
+		if wrote, ok := wroteDBs[desc.GetParentID()]; ok {
+			// If we're creating a new database in this restore, the privileges of the
+			// table and schema should be that of the parent DB.
+			//
+			// Leave the privileges of the temp system tables as the default too.
+			// XXX(pbardea): I deleted `|| wrote.GetName() == restoreTempSystemDB`. Is this ok?j
```

Comment:
I still have to look into if this condition was required, but wanted to get the PR out sooner rather than later to test against CI for any unexpected failures.